### PR TITLE
fix(record): add synchronized so that record creation is sync

### LINKED ISSUE #79: getRecord() does not seem to finish
In my initial Activity I set up Deepstreem as shown below: 

```java
DeepstreamFactory factory = DeepstreamFactory.getInstance();
try {
    deepstreamClient = factory.getClient("wss://...");
    JsonElement auth = new JsonParser().parse("{username: Safinn}");
    deepstreamClient.login(auth);
} catch (URISyntaxException uriSyntaxException) {
    // Deal with error
}
```

In another activity I try and use Deepstream:

```java
DeepstreamFactory deepstreamFactory = DeepstreamFactory.getInstance();
deepstreamClient = deepstreamFactory.getClient();

new Thread(new Runnable() {
    @Override
    public void run() {
        Record userRecord = deepstreamClient.record.getRecord("user");
        userRecord.set("room", "topic");
        userRecord.discard();
    }
}).start();
```

My problem is that it never seems to get past the getRecord() line and the data does not show up in Deepstream hub. It just seems to hang on that line. @AlexBHarley

### PR DESCRIPTION
**Pull request by AlexBHarley**
Fixes #79

### CODE REVIEW

File: src/main/java/io/deepstream/Record.java [hunk 1]
```diff
@@ -509,7 +509,9 @@ Record whenReady(RecordReadyListener recordReadyListener) {
         if( this.isReady ) {
             recordReadyListener.onRecordReady( this.name, this );
         } else {
-            this.onceRecordReadyListeners.add( recordReadyListener );
+            synchronized (this) {
```
Comment (yasserf) [inline src/main/java/io/deepstream/Record.java]: why is this needed?

File: src/main/java/io/deepstream/RecordHandler.java [hunk 2]
```diff
@@ -59,13 +59,17 @@ public Record getRecord( String name ) {
         Record record = records.get( name );
         if( record == null ) {
             synchronized (this) {
```
Comment (yasserf) [inline src/main/java/io/deepstream/RecordHandler.java]: might be worth pulling this into a seperate private function

File: src/main/java/io/deepstream/Record.java (hunk 1 above)
Comment (AlexBHarley) [inline src/main/java/io/deepstream/Record.java]: Without this all record ready listeners are not added properly. So when the read comes back  some are null. By adding them sync, they can all get added

Comment (yasserf) [inline src/main/java/io/deepstream/Record.java]: Still don't get it tbh :$

Comment (AlexBHarley) [inline src/main/java/io/deepstream/Record.java]: I'm not 100% sure either, the problem is here.

```
private void onRead( Message message ) {
    ackTimeoutRegistry.clear( message );

    Map<String,JsonElement> oldValues = beginChange();
    this.version = Integer.parseInt( message.data[ 1 ] );
    this.data = gson.fromJson( message.data[ 2 ], JsonElement.class );
    this.path.setCoreElement(this.data);
    completeChange( oldValues );
    setReady();
}

private void setReady() {
    this.isReady = true;

    for(RecordReadyListener recordReadyListener: this.onceRecordReadyListeners) {
        recordReadyListener.onRecordReady( this.name, this );
    }
    this.onceRecordReadyListeners.clear();
}
```

When there are multiple threads with a reference to a record, often the `recordReadyListener` is null. Causing a NPE.

By putting synchronized, this never happens.

**Comment by yasserf** [comment]
I'm okay with this.

File: src/main/java/io/deepstream/Record.java (hunk 1 above)
Comment (honorcode) [inline src/main/java/io/deepstream/Record.java]: FYI - I re-ran my tests, both versions (one client shared by all threads, one client per thread) passed with no exceptions running locally with 200 threads.

**Comment by honorcode** [comment]
FYI - I re-ran my tests, both versions (one client shared by all threads, one client per thread) passed with no exceptions running locally with 200 threads.

**Comment by AlexBHarley** [comment]
thanks @honorcode, we'll look to get this release tomorrow.

**Comment by yasserf** [comment]
@AlexBHarley why be tests always failing btw? =D

**Comment by yasserf** [comment]
thank you @honorcode for qa'n this!

**Comment by AlexBHarley** [comment]
I think the build-prod.gradle is being run every time. So it will only pass when there is a version increment and a push to master. Will look into it at some point

**Pull request merged**